libs/utils/include/utils/darwin/Systrace.h

```diff
@@ -0,0 +1,259 @@
+        if (tag && UTILS_UNLIKELY(mIsTracingEnabled)) {
+            char buf[64];
+            snprintf(buf, 64, "%s - %d", name, value);
+            APPLE_SIGNPOST_EMIT(sGlobalState.systraceLog, OS_SIGNPOST_EVENT,
+                    OS_SIGNPOST_ID_EXCLUSIVE, name, buf)
+        }
+    }
+
+    inline void value(uint32_t tag, const char* name, int64_t value) noexcept {
+        if (tag && UTILS_UNLIKELY(mIsTracingEnabled)) {
+            char buf[64];
+            snprintf(buf, 64, "%s - %lld", name, value);
+            APPLE_SIGNPOST_EMIT(sGlobalState.systraceLog, OS_SIGNPOST_EVENT,
+                    OS_SIGNPOST_ID_EXCLUSIVE, name, buf)
+        }
+    }
+
+    inline void frameId(uint32_t tag, uint32_t frame) noexcept {
+        if (tag && UTILS_UNLIKELY(mIsTracingEnabled)) {
+            char buf[64]; \
+            snprintf(buf, 64, "frame %u", frame); \
+            APPLE_SIGNPOST_EMIT(sGlobalState.frameIdLog, OS_SIGNPOST_EVENT,
+                    OS_SIGNPOST_ID_EXCLUSIVE, "frame", buf)
+        }
+    }
+
+   private:
+    friend class ScopedTrace;
+
+    struct GlobalState {
+        std::atomic<uint32_t> isTracingEnabled;
+
+        os_log_t systraceLog;
+        os_log_t frameIdLog;
+    };
+
+    static GlobalState sGlobalState;
+
+    void init(uint32_t tag) noexcept;
+
+    // cached values for faster access, no need to be initialized
+    bool mIsTracingEnabled;
+
+    static void setup() noexcept;
+    static void init_once() noexcept;
+    static bool isTracingEnabled(uint32_t tag) noexcept;
+};
+
+// ------------------------------------------------------------------------------------------------
+
+class ScopedTrace {
+   public:
+    // we don't inline this because it's relatively heavy due to a global check
+    ScopedTrace(uint32_t tag, const char* name) noexcept : mTrace(tag), mName(name), mTag(tag) {
+        mTrace.traceBegin(tag, name);
+    }
+
+    inline ~ScopedTrace() noexcept {
+        mTrace.traceEnd(mTag, mName);
+    }
+
+    inline void value(uint32_t tag, const char* name, int32_t v) noexcept {
+        mTrace.value(tag, name, v);
+    }
+
+    inline void value(uint32_t tag, const char* name, int64_t v) noexcept {
+        mTrace.value(tag, name, v);
+    }
+
+   private:
+    Systrace mTrace;
+    const char* mName;
+    const uint32_t mTag;
+};
+
+} // namespace details
+} // namespace utils
+
+// ------------------------------------------------------------------------------------------------
+#else // !ANDROID && !__APPLE__
+// ------------------------------------------------------------------------------------------------
+
+#define SYSTRACE_ENABLE()
+#define SYSTRACE_DISABLE()
+#define SYSTRACE_CONTEXT()
+#define SYSTRACE_NAME(name)
+#define SYSTRACE_NAME_BEGIN(name)
+#define SYSTRACE_NAME_END()
+#define SYSTRACE_CALL()
+#define SYSTRACE_ASYNC_BEGIN(name, cookie)
+#define SYSTRACE_ASYNC_END(name, cookie)
+#define SYSTRACE_VALUE32(name, val)
+#define SYSTRACE_VALUE64(name, val)
```

Comment:
Seems like this is covered in `utils/Systrace.h`?

Reply:
Yes, good catch.